src/wallet/silentpayments.h

```diff
@@ -0,0 +1,178 @@
+ * change outputs for the receiver. Additional labels can be specified starting from `m=1`. The counter m
+ * is hashed with the secret scan key to prevent an external observer from learning anything about the
+ * receiver's labels.
+ *
+ * It is strongly recommended labels be created sequentially starting from `m=1`. Doing so ensures that
+ * label data is not critical for backups, as the receiving wallet can pregenerate labels starting from 1
+ * up to some sufficiently large M when recovering from a backup.
+ *
+ * @param scan_key The secret scan key.
+ * @param m Integer representing the "m-th" label. m=0 is reserved for the change output.
+ * @return uint256 The computed silent payment label tweak.
+ */
+uint256 ComputeSilentPaymentLabelTweak(const CKey& scan_key, const int m);
+
+/**
+ * @brief Generate a silent payment labeled address.
+ *
+ * @param receiver The receiver's silent payment destination (i.e. scan and spend public keys).
+ * @param label The label tweak
+ * @return V0SilentPaymentDestination The slient payment destination, with `B_spend -> B_spend + label * G`.
+ *
+ * @see ComputeSilentPaymentLabelTweak(const CKey& scan_key, const int m);
+ */
+V0SilentPaymentDestination GenerateSilentPaymentLabeledAddress(const V0SilentPaymentDestination& receiver, const uint256& label);
+
+/**
+ * @brief Generate silent payment taproot destinations.
+ *
+ * Given a set of silent payment destinations, generate the requested number of outputs. If a silent payment
+ * destination is repeated, this indicates multiple outputs are requested for the same recipient. The silent payment
+ * desintaions are passed in map where the key indicates their desired position in the final tx.vout array.
+ *
+ * @param seckey      The private key. This private key is either the receiver's scan key or the
+ *                    sum of the sender's input private keys (excluding inputs not eligible for
+ *                    shared secret derivation).
+ * @param inputs_hash The tagged hash of hash of the smallest outpoint and sum of the eligible
+ *                    input public keys.
+ * @param sp_dests    The silent payment destinations.
+ * @return std::map<size_t, WitnessV1Taproot> The generated silent payment taproot destinations.
+ *
+ * @see ComputeInputsHash(const std::vector<COutPoint>& tx_outpoints, const CPubKey& sum_input_pubkeys);
+ */
+std::map<size_t, WitnessV1Taproot> GenerateSilentPaymentTaprootDestinations(const CKey& seckey, const uint256& inputs_hash, const std::map<size_t, V0SilentPaymentDestination>& sp_dests);
+
+/**
+ * @brief Get silent payment tweak data from transaction inputs.
+ *
+ * Get the necessary data from the transaction inputs to be able to scan the transaction outputs for silent payment outputs.
+ * This requires knowledge of the prevout scriptPubKey, which is passed via `coins`.
+ *
+ * If the transaction has silent payment eligible inputs, the public keys from these inputs are returned as a sum along with
```

Comment:
I'm too dumb to understand why do we need to return sum on input pubkeys together with the input hash. Does it have something to do with index and light clients?

Reply:
Yep! There are two ways to use this data:

1. Immediately, as `(b_scan * input_hash) * sum_input_pubkeys`. This is one scalar multiplication and one EC multiplication.
2. Store it for later (either for wallet re-scans or to serve to light clients), as `input_hash * sum_input_pubkeys`. This is one EC multiplication and allows us to store all of the input data needed as 33 bytes.

If the function instead returned `input_hash * sum_input_pubkeys`, we would end up doing two EC multiplications for case 1 (i.e. `(input_hash * sum_input_pubkeys) * b_scan`)